### PR TITLE
More node stats

### PR DESCRIPTION
**Pull request by andrewvc**
This resolves the non-plugin portion of #4923 

Fixes https://github.com/elastic/logstash/issues/5475

It also DRYs up some of the metrics/API code.

### CODE REVIEW

**Comment by suyograo** [comment]
LGTM, looks straightforward

File: logstash-core/lib/logstash/agent.rb [hunk 1]
```diff
@@ -47,13 +47,14 @@ def initialize(settings = LogStash::SETTINGS)
     @collect_metric = setting("metric.collect")
     @metric = create_metric_collector
     @periodic_pollers = LogStash::Instrument::PeriodicPollers.new(create_metric_collector)
+    
+    start_background_services
```
Comment (andrewvc) [inline logstash-core/lib/logstash/agent.rb]: This was put here to address #5366 but doesn't fix it completely.

File: logstash-core/lib/logstash/api/commands/node.rb [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+            :mem => {
+              :heap_init_in_bytes => (memory_bean.getHeapMemoryUsage().getInit() < 0 ? 0 : memory_bean.getHeapMemoryUsage().getInit()),
+              :heap_max_in_bytes => (memory_bean.getHeapMemoryUsage().getMax() < 0 ? 0 : memory_bean.getHeapMemoryUsage().getMax()),
+              :non_heap_init_in_bytes => (memory_bean.getNonHeapMemoryUsage().getInit() < 0 ? 0 : memory_bean.getNonHeapMemoryUsage().getInit()),
+              :non_heap_max_in_bytes => (memory_bean.getNonHeapMemoryUsage().getMax() < 0 ? 0 : memory_bean.getNonHeapMemoryUsage().getMax())
+            }
+          }
+        end
+
+        def hot_threads(options={})
```
Comment (andrewvc) [inline logstash-core/lib/logstash/api/commands/node.rb]: This was moved from Stats because it was never a Stats API call. This is its proper home.

File: logstash-core/lib/logstash/api/commands/stats.rb [hunk 3]
```diff
@@ -5,17 +5,32 @@ module LogStash
   module Api
     module Commands
       class Stats < Commands::Base
-
         def jvm
-          {:threads => service.get_shallow(:jvm, :threads)}
+          {
```
Comment (andrewvc) [inline logstash-core/lib/logstash/api/commands/stats.rb]: This was refactored to use `extract_metrics` to make the mapping of metrics to field names more explicit. This way we can add metrics internally to a namespace without necessarily exposing them to the API.

File: logstash-core/lib/logstash/api/commands/stats.rb [hunk 4]
```diff
@@ -34,70 +49,6 @@ def memory
               acc
             end
           }
-        end
-
-        def hot_threads(options={})
```
Comment (andrewvc) [inline logstash-core/lib/logstash/api/commands/stats.rb]: Moved to the nodes API where it should have been all along.

**Comment by andrewvc** [comment]
@purbon @ph some comment would be welcome here! We won't get passing tests till we address #5366

File: logstash-core/lib/logstash/api/commands/base.rb [hunk 5]
```diff
@@ -19,6 +19,43 @@ def uptime
...
+        #   [:cpu, [:total_in_millis, :percent]]
+        # )
+        # 
+        # Returns:
+        # # From the jvm.process metrics namespace
+        # {
+        #   :open_file_descriptors => 123
+        #   :cpu => { :total_in_millis => 456, :percent => 789 }
+        # }
+        def extract_metrics(path, *keys)
```
Comment (purbon) [inline logstash-core/lib/logstash/api/commands/base.rb]: might this function something to be useful for other consumers of the api?

File: logstash-core/lib/logstash/api/commands/node.rb [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+          {
+            :pid =>  ManagementFactory.getRuntimeMXBean().getName().split("@").first.to_i,
+            :version => java.lang.System.getProperty("java.version"),
+            :vm_name => java.lang.System.getProperty("java.vm.name"),
+            :vm_version => java.lang.System.getProperty("java.version"),
+            :vm_vendor => java.lang.System.getProperty("java.vendor"),
+            :vm_name => java.lang.System.getProperty("java.vm.name"),            
+            :start_time_in_millis => started_at,
+            :mem => {
+              :heap_init_in_bytes => (memory_bean.getHeapMemoryUsage().getInit() < 0 ? 0 : memory_bean.getHeapMemoryUsage().getInit()),
```
Comment (purbon) [inline logstash-core/lib/logstash/api/commands/node.rb]: can we make all this super long lines shorted and easier to digest? I think we should be avoiding ternary like this as longer as the line it gets more hard to follow the logic.

File: logstash-core/spec/api/lib/api/support/resource_dsl_methods.rb [hunk 7]
```diff
@@ -14,7 +14,13 @@ def hash_to_mapping(h, path=[], mapping={})
 
   def test_api(expected, path)
     context "GET #{path}" do
-      let(:payload) { LogStash::Json.load(last_response.body) }
+      let(:payload) do
+        begin
+          LogStash::Json.load(last_response.body)
+        rescue LogStash::Json::ParserError => e
+          require 'pry'; binding.pry
```
Comment (purbon) [inline logstash-core/spec/api/lib/api/support/resource_dsl_methods.rb]: pry when something wrong happens? I think this might be good for debugging but can cause unnecessary problems if raised unintentional

File: logstash-core/lib/logstash/api/rack_app.rb [hunk 8]
```diff
@@ -90,6 +90,8 @@ def self.app(logger, environment)
           run LogStash::Api::Modules::Root
           namespaces.each_pair do |namespace, app|
             map(namespace) do
+
```
Comment (purbon) [inline logstash-core/lib/logstash/api/rack_app.rb]: empty?

File: logstash-core/lib/logstash/api/commands/base.rb (hunk 5 above)
Comment (andrewvc) [inline logstash-core/lib/logstash/api/commands/base.rb]: Yes. @ph proposed moving this into the collector. I agree with him on this. I'll wait till #5366 is addressed before I attempt that move.

File: logstash-core/lib/logstash/api/commands/node.rb (hunk 6 above)
Comment (andrewvc) [inline logstash-core/lib/logstash/api/commands/node.rb]: I agree that it isn't so great to read, but I think its better than splitting into a separate line where it takes up less horizontal but more vertical space.

I'm basically stealing this from https://github.com/elastic/elasticsearch/blob/34da01f794ba892176376aafc18e901cacdd0fce/core/src/main/java/org/elasticsearch/monitor/jvm/JvmInfo.java#L72

I feel like the ES team probably made the same tradeoff

File: logstash-core/lib/logstash/api/rack_app.rb (hunk 8 above)
Comment (andrewvc) [inline logstash-core/lib/logstash/api/rack_app.rb]: good catch!

File: logstash-core/spec/api/lib/api/support/resource_dsl_methods.rb (hunk 7 above)
Comment (andrewvc) [inline logstash-core/spec/api/lib/api/support/resource_dsl_methods.rb]: That is a bug! I def need to remove this :). Thanks for finding this.

**Comment by andrewvc** [comment]
@ph @purbon ready for another round of review!

**Comment by suyograo** [comment]
@andrewvc agreed that `hot_threads` should be under `_node`, but I wonder if we should omit it when we call `GET curl localhost:9600/_node?pretty` and only show it when someone explicitly calls `GET curl localhost:9600/_node/hot_threads?pretty`

**Comment by suyograo** [comment]
Also, would you mind moving `_plugins` from `localhost:9600/_plugins` to `localhost:9600/_node/plugins` as noted in #5475 ?

**Comment by suyograo** [comment]
Tested this PR, info looks good.

```
suyog@machine:~$ curl localhost:9600/_node?pretty
{
  "pipeline" : {
    "workers" : 8,
    "batch_size" : 125,
    "batch_delay" : 5
  },
  "os" : {
    "name" : "Mac OS X",
    "arch" : "x86_64",
    "version" : "10.11.1",
    "available_processors" : 8
  },
  "jvm" : {
    "pid" : 92317,
    "version" : "1.8.0_65",
    "vm_name" : "Java HotSpot(TM) 64-Bit Server VM",
    "vm_version" : "1.8.0_65",
    "vm_vendor" : "Oracle Corporation",
    "start_time_in_millis" : 1465867138764,
    "mem" : {
      "heap_init_in_bytes" : 268435456,
      "heap_max_in_bytes" : 1037959168,
      "non_heap_init_in_bytes" : 2555904,
      "non_heap_max_in_bytes" : 0
    }
  },
```

File: logstash-core/lib/logstash/instrument/metric_store.rb [hunk 9]
```diff
@@ -110,6 +110,54 @@ def get_shallow(*key_paths)
...
+          # This is a Metric object, we need to extract its value
+          # If the metric didn't exist it might be nil, but we still want its key
+          # to exist with a nil value
+          acc[k] = v ? v.value : nil; acc
+        end
+        
+        acc
+      end
+    end    
+
```
Comment (ph) [inline logstash-core/lib/logstash/instrument/metric_store.rb]: Thanks for moving this inside the metric store

Comment (ph) [inline logstash-core/lib/logstash/instrument/metric_store.rb]: Can we add a unit test for it?

**Comment by ph** [comment]
Tested this PR, info looks goods.
Only a minor comments concerning `extract_metrics`, I think we should have test covering it, its a recursive method and I would be more confident if we have test targetting it.

**Comment by andrewvc** [comment]
@ph I've added extensive unit tests to `extract_metrics` and fully refactored it to make it act correctly with deep metrics and weird stuff like `[:foo, [:pipeline1, :pipeline2], :common]`

File: logstash-core/lib/logstash/instrument/metric_store.rb (hunk 9 above)
Comment (andrewvc) [inline logstash-core/lib/logstash/instrument/metric_store.rb]: Done!

**Comment by andrewvc** [comment]
@suyograo @ph I believe I've made all required changes. How's it looking?

**Comment by suyograo** [comment]
Tested, LGTM from my end.

**Comment by andrewvc** [comment]
@ph can I get an LGTM from you to proceed

**Comment by ph** [comment]
Tested L G T M
